### PR TITLE
Expense attachments

### PR DESCRIPTION
**Pull request by chatterjeeshekhar**
Fetch the attachments from Fyle and push it to QuickBooks Online. Supported expenses: JournalEntry, CC Purchase, Cheque, Bills

### CODE REVIEW

File: run.sh [hunk 1]
```diff
@@ -5,7 +5,7 @@
 source setup.sh
 
 # Run db migrations
-python manage.py migrate
+python3 manage.py migrate
```
Comment (Shwetabhk) [inline run.sh:8]: change it back to python

File: apps/fyle/utils.py [hunk 2]
```diff
@@ -62,3 +62,19 @@ def get_projects(self, active_only: bool):
         Get projects from fyle
         """
         return self.connection.Projects.get(active_only=active_only)['data']
+
+    def get_attachments(self, expense_ids: List[str]):
```
Comment (Shwetabhk) [inline apps/fyle/utils.py:66]: There is a merge conflict in this file, please pull the latest master, merge it into your private branch and resolve it

File: apps/quickbooks_online/tasks.py [hunk 3]
```diff
@@ -24,6 +24,55 @@
...
+            attachment = fyle_sdk_connection.Expenses.get_attachments(expense_id)
+            if attachment['data']:
+                attachment = attachment['data'][0]
+                attachment['expense_id'] = expense_id
+                attachments.append(attachment)
+        return attachments
+
+    return []
+
+def load_attachments(connection, ref_id: str, ref_type: str, attachments) -> List:
```
[the remark this answers is not in the document]
Comment (chatterjeeshekhar) [reply apps/quickbooks_online/tasks.py:52]: Like replicate it, or move it and call the function from there?

File: apps/quickbooks_online/tasks.py [hunk 4]
```diff
@@ -24,6 +24,55 @@
...
+        :param ref_type: type of object
+        :return: True for success, False for failure
+        """
+
+        logger.info('Loading attachments to QBO')
+
+        if len(attachments):
+            responses = []
+            for attachment in attachments:
+                response = connection.attachments.post(
```
[the remark this answers is not in the document]
Comment (chatterjeeshekhar) [reply apps/quickbooks_online/tasks.py:66]: Okay, ignore the other question. Clear now.

File: apps/quickbooks_online/tasks.py [hunk 5]
```diff
@@ -345,6 +411,14 @@ def create_credit_card_purchase(expense_group, task_log):
                 credit_card_purchase_object, credit_card_purchase_lineitems_objects
             )
 
+            #Load Attachment
+            expense_ids = expense_group.expenses.values_list('expense_id', flat=True)
+
+            attachments = get_expense_id_attachments_group(expense_group.workspace_id, expense_ids)
+            
+            load_attachments(qbo_connection.connection, created_credit_card_purchase["Purchase"]["Id"], "Purchase", attachments)
```
[the remark this answers is not in the document]
Comment (chatterjeeshekhar) [reply apps/quickbooks_online/tasks.py:419]: what will be in except?

[the remark this answers is not in the document]
Comment (chatterjeeshekhar) [reply apps/quickbooks_online/tasks.py:419]: Got it from the other comment

File: apps/quickbooks_online/tasks.py [hunk 6]
```diff
@@ -84,6 +133,15 @@ def create_bill(expense_group, task_log):
 
             created_bill = qbo_connection.post_bill(bill_object, bill_lineitems_objects)
 
+            #Load Attachment
+            expense_ids = expense_group.expenses.values_list('expense_id', flat=True)
+
+            attachments = get_expense_id_attachments_group(expense_group.workspace_id, expense_ids)
+            print(created_bill)
+
+            load_attachments(qbo_connection.connection, created_bill["Bill"]["Id"], "Purchase", attachments)
```
[the remark this answers is not in the document]
Comment (chatterjeeshekhar) [reply apps/quickbooks_online/tasks.py:142]: Done!

[the remark this answers is not in the document]
Comment (chatterjeeshekhar) [reply apps/quickbooks_online/tasks.py:142]: from .utils import QBOConnector

File: apps/quickbooks_online/utils.py [hunk 7]
```diff
@@ -335,3 +335,27 @@ def post_journal_entry(self, journal_entry: JournalEntry, journal_entry_lineitem
...
+        :param ref_type: type of object
+        :return: True for success, False for failure
+        """
+
+        logger.info('Loading attachments to QBO')
+
+        if len(attachments):
+            responses = []
+            for attachment in attachments:
+                response = self.attachments.post(
```
Comment (Shwetabhk) [inline apps/quickbooks_online/utils.py:353]: this should be `self.connection.attachments.post`

File: setup_template.sh [hunk 8]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 # Django Settings
-export SECRET_KEY=YOUR DJANGO SECRET KEY
+export SECRET_KEY=YOUR DJANGO SECRET KEY
```
Comment (Shwetabhk) [inline setup_template.sh:4]: remove extra space

File: apps/quickbooks_online/utils.py [hunk 9]
```diff
@@ -335,3 +335,27 @@ def post_journal_entry(self, journal_entry: JournalEntry, journal_entry_lineitem
...
+    def load_attachments(self, ref_id: str, ref_type: str, attachments) -> List:
+        """
+        Link attachments to objects Quickbooks
+        :param prep_id: prep id for export
+        :param ref_id: object id
+        :param ref_type: type of object
+        :return: True for success, False for failure
+        """
+
+        logger.info('Loading attachments to QBO')
```
Comment (Shwetabhk) [inline apps/quickbooks_online/utils.py:348]: remove logger

File: apps/quickbooks_online/utils.py [hunk 10]
```diff
@@ -335,3 +335,27 @@ def post_journal_entry(self, journal_entry: JournalEntry, journal_entry_lineitem
         journal_entry_payload = self.__construct_journal_entry(journal_entry, journal_entry_lineitems)
         created_journal_entry = self.connection.journal_entries.post(journal_entry_payload)
         return created_journal_entry
+
+    def load_attachments(self, ref_id: str, ref_type: str, attachments) -> List:
```
Comment (Shwetabhk) [inline apps/quickbooks_online/utils.py:339]: rename to post_attachmebts

File: apps/quickbooks_online/utils.py [hunk 11]
```diff
@@ -335,3 +335,27 @@ def post_journal_entry(self, journal_entry: JournalEntry, journal_entry_lineitem
...
+            for attachment in attachments:
+                response = self.attachments.post(
+                    ref_id=ref_id,
+                    ref_type=ref_type,
+                    content=attachment['content'],
+                    file_name=attachment['filename']
+                )
+                responses.append(response)
+            return responses
+        return []
```
Comment (Shwetabhk) [inline apps/quickbooks_online/utils.py:361]: add a new line at the end of the file

File: apps/fyle/utils.py [hunk 12]
```diff
@@ -39,26 +43,103 @@ def get_expenses(self, state: List[str], updated_at: List[str], fund_source: Lis
...
+        if expense_ids:
+            for expense_id in expense_ids:
+                attachment = self.Expenses.get_attachments(expense_id)
+                if attachment['data']:
+                    attachment = attachment['data'][0]
+                    attachment['expense_id'] = expense_id
+                    attachments.append(attachment)
+            return attachments
+
+        return []
```
Comment (Shwetabhk) [inline apps/fyle/utils.py:145]: New line at the end of the file

File: apps/fyle/views.py [hunk 13]
```diff
@@ -19,7 +19,7 @@ class ExpenseGroupView(generics.ListCreateAPIView):
     List Fyle Expenses
     """
     serializer_class = ExpenseGroupSerializer
-
+
```
Comment (Shwetabhk) [inline apps/fyle/views.py:22]: remove spaces on line 22

File: apps/quickbooks_online/tasks.py [hunk 14]
```diff
@@ -355,6 +379,9 @@ def create_credit_card_purchase(expense_group, task_log):
             created_credit_card_purchase = qbo_connection.post_credit_card_purchase(
                 credit_card_purchase_object, credit_card_purchase_lineitems_objects
             )
+            
+            load_attachments(qbo_connection, created_credit_card_purchase['Purchase']['Id'], 'Purchase', expense_group)
+
```
Comment (Shwetabhk) [inline apps/quickbooks_online/tasks.py:384]: remove extra empty lines, keep a maximum gap of 1

File: apps/quickbooks_online/tasks.py [hunk 15]
```diff
@@ -27,6 +27,26 @@
...
+    :param ref_type: type of object
+    :param expense_group: Expense group
+    """
+    try:
+        fyle_credentials = FyleCredential.objects.get(workspace_id=expense_group.workspace_id)
+        expense_ids = expense_group.expenses.values_list('expense_id', flat=True)
+        fyle_connector = FyleConnector(fyle_credentials, expense_group.workspace_id)
+        attachments = fyle_connector.get_attachments(expense_ids)
+        qbo_connection.post_attachments(ref_id, ref_type, attachments)
+    except Exception as e:
```
Comment (Shwetabhk) [inline apps/quickbooks_online/tasks.py:44]: since `e` is not used anywhere
replace `Exception as e:` to `Exception:`

**Pull request merged**